### PR TITLE
Metadata file import service

### LINKED ISSUE #356: Project Sample Metadata - File Import Service
## Summary

As a user I want to be able to import metadata for a batch of samples from a file. csv, excel files should be supported. I want the option to ignore empty values, which would mean the metadata fields with empty values are not deleted from the sample metadata

## Acceptance Criteria

- Create a `Samples::Metadata::FileImportService`
- The arguments are going to be the `project`, `user`, params is a hash which contains the following keys: `file` which is a ruby file object, `sample_id_column`, `ignore_empty_values` 
- The service will validate that user can modify the project
- The file is valid (contains atleast the sample_id column, atleast one metadata column, and atleast one row of sample metadata)
- The service should utilize the `Samples::Metadata::UpdateService` to update the metadata
- If `ignore_empty_values` is set, we want to modify the payload sent to the update service to strip the fields which have empty values
- Service should build a response with the response received from the update service for each update

### PR DESCRIPTION
**Pull request by ksierks**
## What does this PR do and why?
This service imports metadata for a batch of samples from a CSV or Excel file. It has the option to ignore empty values. Every response from the update service will be collected and returned when the file is finished processing.
Fixes #356.

## Screenshots or screen recordings
N/A

## How to set up and validate locally
- Test via running the service tests:
  1. Run tests within `test/services/samples/metadata/file_import_service_test.rb`
- Test via console:
  1. Open a console
  2. Run the command: `Samples::Metadata::FileImportService.new(project, user, { file:, sample_id_column:, ignore_empty_values: }).execute`

Note: Example files can be found in the `test/fixtures/files/metadata` folder.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/services/samples/metadata/file_import_service.rb
Comment (ChrisHuynh333) [inline app/services/samples/metadata/file_import_service.rb:95]: `status` can be renamed as the `UpdateService` in [PR 339](https://github.com/phac-nml/irida-next/pull/339) (not yet in `main`) has been updated to return `@metadata_changes = { added: [], updated: [], deleted: [], not_updated: [] }`, where updated keys are added to the appropriate action, and is no longer called `update_status`.

Also I believe `@current_user` can just be `current_user`

File: app/services/samples/metadata/file_import_service.rb
Comment (ChrisHuynh333) [inline app/services/samples/metadata/file_import_service.rb:28]: On line 98, you're returning `response`, so I'm assuming you'd want to assign it to a variable here, and then presumably also return it to the controller (unless you're waiting to plug it into a controller first).

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ChrisHuynh333) [inline test/services/samples/metadata/file_import_service_test.rb:45]: These will need to be updated throughout the testing once [PR 339](https://github.com/phac-nml/irida-next/pull/339) is merged with the new hash values as mentioned in the previous comment.

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ChrisHuynh333) [inline test/services/samples/metadata/file_import_service_test.rb:112]: Eric mentioned in a PR of mine he wanted me to get away from setting the metadata like so in a test and to use fixtures with metadata already added.
[PR339](https://github.com/phac-nml/irida-next/pull/339) also contains a bunch of new fixtures that contains metadata.

@ericenns did you want @ksierks to utilize these fixtures once merged into `main`?

**Review by ChrisHuynh333 — CHANGES_REQUESTED** [review]
Looks great so far.
I've left some comments.
I would recommend putting this PR On Hold and wait for [PR 339](https://github.com/phac-nml/irida-next/pull/339) to be merged in as most of my comments are related to changes contained in 339.
@ericenns @deepsidhu85 thoughts?

**Comment by ericenns** [comment]
> Looks great so far. I've left some comments. I would recommend putting this PR On Hold and wait for [PR 339](https://github.com/phac-nml/irida-next/pull/339) to be merged in as most of my comments are related to changes contained in 339. @ericenns @deepsidhu85 thoughts?

No need to put on On Hold, just need to rebase against main and update this PR based on changes introduced in PR 339 as I just merged it in.

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb]: Please see 4685902c953494151b4408fb2b406c8e400d0bbd.

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb]: I removed status all together. Please see b22e4215fb79b2d30969749b79fea6eb77caabfa.

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb:28]: My understanding is that the last line is always returned. The tests are able to evaluate the response. The controller should be able to call the service and receive the response. I think we're okay.

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb]: Please see 8dab7b7fb656212806f88f0332ae792c31a1f2e1.

**Review by ChrisHuynh333 — DISMISSED** [review]
Looks good to me!

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [inline app/services/samples/metadata/file_import_service.rb:16]: We shouldn't need to explicitly set false here. If it doesn't find the param it will be set to nil which we can use

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [inline app/services/samples/metadata/file_import_service.rb:51]: @ericenns do we want to support open office spreadsheet formats?

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [inline app/services/samples/metadata/file_import_service.rb:49]: A little refactoring to possibly remove some of the rubocop disables on this method. We can move the file extension validation into it's own method, the validation of the file contents (opening and checking for the columns) can be moved into it's own method as well

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [inline app/services/samples/metadata/file_import_service.rb:70]: I think it would be a good idea to check for duplicate headers and raise a error advising the user that there are duplicate headers and to fix the file and re-try

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [inline app/services/samples/metadata/file_import_service.rb:86]: Let's rename this variable to sample_id_column so it's clear what it is

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This is looking good just have a few comments below. I still need to look over the testing section and manually test

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb]: 0455419f8ab52f8a4ce8956eda4a29c21195e97b

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb]: 0455419f8ab52f8a4ce8956eda4a29c21195e97b

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb:44]: 0455419f8ab52f8a4ce8956eda4a29c21195e97b

File: app/services/samples/metadata/file_import_service.rb
Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb]: 0455419f8ab52f8a4ce8956eda4a29c21195e97b

File: test/services/samples/metadata/file_import_service_test.rb
Comment (deepsidhu85) [inline test/services/samples/metadata/file_import_service_test.rb:50]: Can you add some assert statements above the call to the file import service which will assert_equal what the metadata is for the sample before import

File: test/services/samples/metadata/file_import_service_test.rb
Comment (deepsidhu85) [inline test/services/samples/metadata/file_import_service_test.rb:65]: Same comment as above

File: test/services/samples/metadata/file_import_service_test.rb
Comment (deepsidhu85) [inline test/services/samples/metadata/file_import_service_test.rb:79]: Same comment as above

File: test/services/samples/metadata/file_import_service_test.rb
Comment (deepsidhu85) [inline test/services/samples/metadata/file_import_service_test.rb:147]: Same comment as above

File: test/services/samples/metadata/file_import_service_test.rb
Comment (deepsidhu85) [inline test/services/samples/metadata/file_import_service_test.rb:156]: Same comment as above

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Just a few comments about asserting the metadata for a sample before calling the import service

File: config/locales/en.yml
Comment (deepsidhu85) [inline config/locales/en.yml:972]: Can we update this to also advise the user that they need to remove the duplicate column headers

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Just a comment about the translation text

File: config/locales/en.yml
Comment (ksierks) [reply config/locales/en.yml:972]: bcffae4ee9a78635022774876ba167180332d528

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb:164]: bcffae4ee9a78635022774876ba167180332d528

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb:153]: bcffae4ee9a78635022774876ba167180332d528

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb:83]: bcffae4ee9a78635022774876ba167180332d528

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb:67]: bcffae4ee9a78635022774876ba167180332d528

File: test/services/samples/metadata/file_import_service_test.rb
Comment (ksierks) [reply test/services/samples/metadata/file_import_service_test.rb:50]: bcffae4ee9a78635022774876ba167180332d528

File: config/locales/en.yml
Comment (ksierks) [reply config/locales/en.yml]: 9ffc2302f851ae1b9488a395421234d63e38ef98

File: app/services/samples/metadata/file_import_service.rb
Comment (ericenns) [reply app/services/samples/metadata/file_import_service.rb]: I don't think that is necessary, as long as we are supporting `csv`. A person using open office can just save their file as `csv`.

File: app/services/samples/metadata/file_import_service.rb
Comment (ericenns) [inline app/services/samples/metadata/file_import_service.rb:101]: **nitpick (blocking):** can you rename `sample_id_column` variable to `sample_id`. This should be clearer to anyone reading it as we have `@sample_id_column`.

**Review by ericenns — CHANGES_REQUESTED** [review]
This looks good, just a minor comment on variable naming

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [reply app/services/samples/metadata/file_import_service.rb:101]: This was my bad

Comment (ksierks) [reply app/services/samples/metadata/file_import_service.rb]: 271b083be26d2211611a72d444bf0c77f5b3fbe2

**Review by ericenns — APPROVED** [review]

File: app/services/samples/metadata/file_import_service.rb
Comment (deepsidhu85) [reply app/services/samples/metadata/file_import_service.rb]: Sounds good!

**Review by deepsidhu85 — APPROVED** [review]
Thanks for making the changes. Great work!

**Pull request merged**
